### DISCUSSION TITLE
ordering startup systems

### DISCUSSION THREAD

**Discussion by m-mueller678** [bevyengine/bevy#2728] [Q&A]
I am trying to order startup systems so their dependencies are respected (eg. load config files -> load textures from configured locations -> wait for asset loading -> build texture atlas)

I do not want to manually set a fixed sequential order for all startup systems, as that seems tedious and error-prone, and I don't know how to combine this with plugins. 

I tried using `after` and `before`, but this does not work well with run criteria. For example, I am using run criteria to determine if all textures are loaded and the texture atlas can be built, but systems that depend on it will run regardless.

Example: I expected s2 to run only after s1 has run, but that is not the case.
```
fn main() {
    fn s1() {
        println!("s1");
    }

    fn s2() {
        println!("s2");
    }

    App::build()
        // s1 must run before s2 and s2 should be delayed if necessary
        .add_startup_system(s1.system().before("s2").label("s1").with_run_criteria(s1_ready.system()))
        .add_startup_system(s2.system().label("s2"))
        .run();

    fn s1_ready() -> ShouldRun {
        println!("should run?");
        let mut s = String::new();
        std::io::stdin().read_line(&mut s).unwrap();
        if s.trim() == "y" { ShouldRun::Yes } else { ShouldRun::NoAndCheckAgain }
    }
}
```

Another option I considered is creating a State for every startup system that is set to Completed when the system has finished. This would allow running systems after a system they depend on. But this again seems very tedious, and I can't figure out how to create dependencies like `before` using this. For example, a large number of systems might need to load textures before the texture atlas gets built, adding all of them as dependencies to the texture atlas system seems much less clear than adding `before` dependencies to those systems, especially if they are spread across multiple plugins.

Also, both approaches seem to require all initialization to happen in a single frame, which feels like I'm going against how Bevy is meant to be used.

How would I best go about doing this?

**Comment by ManevilleF** [comment]
I would use `State` run criterias.

If you have an state enum like this:

```rust
pub enum AppState {
    Step1,
    Step2, 
}
```

You set the state to `Step1` on building the `App`

You could have your `s2` system declared as:

```
.add_system_set(SystemSet::on_enter(AppState::Step2).with_system(s2.system());
```

And the `s1` system could change the step from `Step1` to `Step2`

**Reply by m-mueller678** [reply]
Unless I'm missing something, this is basically manually setting the order of the systems. This certainly works, I was just hoping, there would be a better approach, maybe using the existing system of `before` and `after` methods for automatic ordering

**Comment by Ratysz** [comment]
Run criteria have no influence on execution order, and execution order has no influence on whether systems run. If `s1` doesn't run, `s2` is still free to do so, and will do so after the point in time where `s1` would normally happen.
